### PR TITLE
✨(models) add edx textbook interaction events pydantic models

### PR DESCRIPTION
**Pull request by quitterie-lcs**
## Purpose

Textbook interaction from `edx` comprises browser events emitted when the user navigates in chapters.

## Proposal

- [x] `pydantic` models
- [x] edge case tests
- [x] `selector` tests

### CODE REVIEW

File: src/ralph/models/edx/textbook_interaction.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+    chapter: constr(
+        regex=(
+            r"^block-v1:[^\/+]+(\/|\+)[^\/+]+(\/|\+)[^\/?]+type"  # noqa : F722
+            r"@sequential\+block@[a-f0-9]{32}.+\.pdf$"  # noqa : F722
+        )
+    )
+
+
+class TextbookPdfThumbnailsToggledEventField(TextbookInteractionBaseEventField):
+    """Representents the `textbook.pdf.thumbnails.toggled` event field.
```
Comment (SergioSim) [inline src/ralph/models/edx/textbook_interaction.py:33]:
```suggestion
    """Represents the `textbook.pdf.thumbnails.toggled` event field.
```
(This seems to be in multiple places)
s/Representents/Represents/

File: src/ralph/models/edx/textbook_interaction.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+        TextbookPdfPageNavigatedEventField,
+    ]
+    event_type: Literal["textbook.pdf.page.navigated"]
+    name: Literal["textbook.pdf.page.navigated"]
+
+
+class UITextbookPdfZoomButtonsChanged(BaseBrowserEvent):
+    """Represents the `textbook.pdf.zoom.buttons.changed` browser event model.
+
+    The browser emits ths event when a user clicks either the <kbd>Zoom In</kbd>
```
Comment (SergioSim) [inline src/ralph/models/edx/textbook_interaction.py:425]:
```suggestion
    The browser emits this event when a user clicks either the <kbd>Zoom In</kbd>
```

**Review by SergioSim — COMMENTED** [review]
This is awesome! I'm impressed) Have only some minor suggestions & questions
image

File: src/ralph/models/edx/textbook_interaction.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+class TextbookPdfDisplayScaledEventField(TextbookInteractionBaseEventField):
+    """Representents the `textbook.pdf.display.scaled` event field.
+
+    Attributes:
+        name (str): Consists of the value `textbook.pdf.display.scaled`.
+        amount (str): Consists of a floating point number string value.
+    """
+
+    name: Literal["textbook.pdf.display.scaled"]
+    amount: constr(regex=r"^[+-]?([0-9]*[.])?[0-9]+$")  # noqa: F722
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/textbook_interaction.py:134]: edx specified it as a [string](https://edx.readthedocs.io/projects/devdata/en/stable/internal_data_formats/tracking_logs.html#textbook-pdf-display-scaled) but I have rechecked and you're right, it's a number

File: src/ralph/models/edx/textbook_interaction.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        query (str): Consists of the value in the search field.
+        status (str): Consists either of the value `not found` for a search string that is
+            unsuccessful or blank for successful search strings.
+    """
+
+    name: Literal["textbook.pdf.search.executed"]
+    caseSensitive: bool
+    highlightAll: bool
+    query: str
+    status: Union[Literal[""], Literal["not found"]]
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/textbook_interaction.py:167]: Ok, so we simplify it only as a `string`?

File: src/ralph/models/edx/textbook_interaction.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+
+    chapter: constr(
+        regex=(
+            r"^block-v1:[^\/+]+(\/|\+)[^\/+]+(\/|\+)[^\/?]+type"  # noqa : F722
+            r"@sequential\+block@[a-f0-9]{32}.+\.pdf$"  # noqa : F722
+        )
+    )
+    name: Union[
+        Literal["textbook.pdf.page.loaded"],
+        Literal["textbook.pdf.page.navigatedprevious"],
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/textbook_interaction.py:255]: We should rather believe it rather than the [documentation](https://edx.readthedocs.io/projects/devdata/en/stable/internal_data_formats/tracking_logs.html#book). Upon reflection, `pdf.navigatedprevious` event does not exist, hence, it would be a non sense

**Review by SergioSim — COMMENTED** [review]
This looks great! Have only some minor suggestions)

File: src/ralph/models/edx/textbook_interaction/fields/events.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+class TextbookPdfThumbnailNavigatedEventField(TextbookInteractionBaseEventField):
+    """Represents the `textbook.pdf.thumbnail.navigated` event field.
+
+    Attribute:
+        name (str): Consists of the value `textbook.pdf.thumbnail.navigated`.
+    """
+
+    name: Literal["textbook.pdf.thumbnail.navigated"]
+
```
Comment (SergioSim) [inline src/ralph/models/edx/textbook_interaction/fields/events.py:46]: thumbnail_title:
```suggestion
    name: Literal["textbook.pdf.thumbnail.navigated"]
    thumbnail_title: str
```

File: src/ralph/models/edx/textbook_interaction/fields/events.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    """
+
+    name: Literal["textbook.pdf.thumbnails.toggled"]
+
+
+class TextbookPdfThumbnailNavigatedEventField(TextbookInteractionBaseEventField):
+    """Represents the `textbook.pdf.thumbnail.navigated` event field.
+
+    Attribute:
+        name (str): Consists of the value `textbook.pdf.thumbnail.navigated`.
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/textbook_interaction/fields/events.py:42]: Good catch! Thanks

File: tests/models/edx/test_textbook_interaction.py [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+    UITextbookPdfZoomMenuChanged,
+)
+from ralph.models.selector import ModelSelector
+
+
+@settings(max_examples=1)
+@given(st.builds(TextbookInteractionBaseEventField))
+def test_fields_edx_textbook_interaction_base_event_field_with_valid_content(field):
+    """Tests that a valid `TextbookInteractionBaseEventField` does not raise
+    a `ValidationError`."""
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply tests/models/edx/test_textbook_interaction.py:51]: You're right. I failed when copy pasting, thanks!

**Comment by quitterie-lcs** [comment]
@SergioSim I have updated with your recommendation :wink:

**Review by SergioSim — APPROVED** [review]
image
